middleware/rate_limiter.go

```diff
@@ -0,0 +1,223 @@
+		DenyHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusForbidden, nil)
+		},
+	}
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiterWithConfig(config))
+*/
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.IdentifierExtractor == nil {
+		config.IdentifierExtractor = DefaultRateLimiterConfig.IdentifierExtractor
+	}
+	if config.ErrorHandler == nil {
+		config.ErrorHandler = DefaultRateLimiterConfig.ErrorHandler
+	}
+	if config.DenyHandler == nil {
+		config.DenyHandler = DefaultRateLimiterConfig.DenyHandler
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier, err := config.IdentifierExtractor(c)
+			if err != nil {
+				return config.DenyHandler(c)
+			}
+
+			if !config.Store.Allow(identifier) {
+				return config.ErrorHandler(c)
+			}
+			return next(c)
+		}
+	}
+}
+
+type (
+	// RateLimiterMemoryStore is the built-in store implementation for RateLimiter
+	RateLimiterMemoryStore struct {
+		visitors    map[string]*Visitor
+		mutex       sync.Mutex
+		rate        rate.Limit
+		burst       int
+		expiresIn   time.Duration
+		lastCleanup time.Time
+	}
+	// Visitor signifies a unique user's limiter details
+	Visitor struct {
+		*rate.Limiter
+		lastSeen time.Time
+	}
+)
+
+// NewRateLimiterMemoryStore returns an instance of RateLimiterMemoryStore
+func NewRateLimiterMemoryStore(config RateLimiterMemoryStoreConfig) (store *RateLimiterMemoryStore) {
+	store = &RateLimiterMemoryStore{}
+
+	store.rate = config.rate
+	store.burst = config.burst
+	if config.expiresIn == 0 {
+		store.expiresIn = DefaultRateLimiterMemoryStoreConfig.expiresIn
+	} else {
+		store.expiresIn = config.expiresIn
+	}
+	store.visitors = make(map[string]*Visitor)
+	store.lastCleanup = now()
+	return
+}
+
+// RateLimiterMemoryStoreConfig represents configuration for RateLimiterMemoryStore
+type RateLimiterMemoryStoreConfig struct {
+	rate      rate.Limit
+	burst     int
+	expiresIn time.Duration
+}
+
+// DefaultRateLimiterMemoryStoreConfig provides default configuration values for RateLimiterMemoryStore
+var DefaultRateLimiterMemoryStoreConfig = RateLimiterMemoryStoreConfig{
+	expiresIn: 3 * time.Minute,
+}
+
+// Allow implements RateLimiterStore.Allow
+func (store *RateLimiterMemoryStore) Allow(identifier string) bool {
+	store.mutex.Lock()
+	limiter, exists := store.visitors[identifier]
+	if !exists {
+		limiter = new(Visitor)
+		limiter.Limiter = rate.NewLimiter(store.rate, store.burst)
+		limiter.lastSeen = now()
+		store.visitors[identifier] = limiter
+	}
+	limiter.lastSeen = now()
+	store.mutex.Unlock()
```

Comment:
I've worked a little on the sharded memory store as an experiment and added benchmark tests
A data race was revealed for high concurrency, so the mutex needs to protect lastCleanup access too, to the `store.mutex.Unlock()` needs to be moved just before the return.

See lammel/echo@17337657676a6dfd7631be364fe9bbf519d1040a
To verify run:

    go test -benchmem -race -run=^$ -bench '^BenchmarkRateLimiter' github.com/labstack/echo/v4/middleware

Reply:
For this PR we will focus on keeping the simple memory store without any risk for races.
Further improvements will be done in a separate PR.

Please merge lammel/echo@f323d36 that will remove the ShardedMemoryStorefor from this PR.